### DISCUSSION TITLE
Documentation for ImageProcessing::Builder?

### DISCUSSION THREAD

**Discussion by brian-nicholls** [janko/image_processing#108] [General]
I am using [CarrierWave::MiniMagick](https://www.rubydoc.info/gems/carrierwave/CarrierWave/MiniMagick). According to the docs:

> "...create your own helpers with the powerful minimagick! method, which yields an ImageProcessing::Builder object."

However I don't see any mention here in the docs of how to use the Builder object. I am trying to composite a PNG watermark onto a PDF.

For reference, the below code using `manipulate!` works on my local environment, but not on the Dockerized production environment. Since the docs say it only exists for backwards compatibility, I am trying to get it working using `minimagick!` via this gem. Any chance someone could help with translating this into a version using `ImageProcessing::Builder`?

```ruby
      manipulate! do |source_pdf|
        watermark_image = MiniMagick::Image.open("#{Rails.root}/app/assets/images/watermark.png")
        source_pdf.composite(watermark_image) do |c|
          c.compose 'overlay'
          c.tile
        end
      end
```

Thanks for any help!

**Comment by janko** [comment]
Anything you see in the [MiniMagick backend docs](https://github.com/janko/image_processing/blob/master/doc/minimagick.md#readme) can be called on the `ImageProcessing::Builder` option. It's just an object that supports chaining, similarly to `ActiveRecord::Relation`.

The following should work:

```rb
minimagick! do |builder|
  watermark_path = "#{Rails.root}/app/assets/images/watermark.png"
  builder.composite(watermark_path, mode: "overlay") do |c|
    c.tile
  end
end
```

**Reply by brian-nicholls** [reply]
Thanks for the quick reply! Unfortunately using the above I am seeing this:
```
CarrierWave::ProcessingError (Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /workspaces/mf_api_server/tmp/1668644236-329297852097376-0005-5183/watermarked/dp_1461_1668644236776.pdf -auto-orient /workspaces/mf_api_server/app/assets/images/watermark.png -compose overlay -tile -composite /tmp/image_processing20221117-18068-1an7xtm.pdf` failed with error:
convert-im6.q16: unable to open image `-composite': No such file or directory @ error/blob.c/OpenBlob/2924.
convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.
):
```
I've seen this error happening using different methods, even when the extension is clearly PDF and should be a supported image. I've checked the delegates file, which does include PDF. I'm stumped, and probably going to have to use a different method using the HexaPDF gem :(

**Reply by janko** [reply]
It doesn't seem the problem is PDF, it's interpreting something else as an image; it says `no decode delegate for this image format ''`, and we see the actual format is missing. Are you sure the `-tile` is valid here? If it is, doesn't `-tile` expect an argument? I think that might be tripping ImageMagick up.

**Reply by brian-nicholls** [reply]
I also tried omitting the `tile` but wound up with a similar error:
```ruby
      minimagick! do |builder|
        watermark_path = "#{Rails.root}/app/assets/images/watermark.png"
        builder.composite(watermark_path, mode: "overlay")
      end
```
Gives me
```
CarrierWave::ProcessingError (Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /workspaces/mf_api_server/tmp/1668696818-126439900497953-0003-9055/watermarked/dp_1484_1668696818715.pdf -auto-orient /workspaces/mf_api_server/app/assets/images/watermark.png -compose overlay -composite /tmp/image_processing20221117-14443-t7rvs9.pdf` failed with error:
convert-im6.q16: image sequence is required `-composite' @ error/mogrify.c/MogrifyImageList/7987.
convert-im6.q16: no images defined `/tmp/image_processing20221117-14443-t7rvs9.pdf' @ error/convert.c/ConvertImageCommand/3229.
):
```
And explicitly returning the builder, gives me another version of the error:
```ruby
      minimagick! do |builder|
        watermark_path = "#{Rails.root}/app/assets/images/watermark.png"
        builder.composite(watermark_path, mode: "overlay")
        builder
      end
```
```
CarrierWave::ProcessingError (Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /workspaces/mf_api_server/tmp/1668697249-551170488041120-0008-3457/watermarked/dp_1489_1668697249843.pdf -auto-orient /tmp/image_processing20221117-14443-1ji0c4p.pdf` failed with error:
convert-im6.q16: no images defined `/tmp/image_processing20221117-14443-1ji0c4p.pdf' @ error/convert.c/ConvertImageCommand/3229.
):
```

What bugs me is that it works perfectly fine on my development environment, but not on the Dockerized staging environment. Because of this I'm convinced it has to have something to do with the filesystem - I can't think of any other difference.
